### PR TITLE
Inference example and inference-framework unit-test for NMT model

### PR DESCRIPTION
**Pull request by sidgoyal78**
Addresses second part of #7999 

*Background*
The fluid example of NMT model has 2 parts:
`train_main()` and `decode_main()`. As discussed with @jacquesqiao , the current implementation of the book chapter (as of Feb 8, 2018), is a bit disjoint. More specifically, the `decode_main` should ideally use the parameters post training (that is, after calling `train_main()`). However, currently the `train_main()` demonstrates the working example of dynamic RNN training, and current `decode_main()` uses a random initialization of the weights in the encoder-decoder and demonstrates correctness of beam search. Hence, both of the modules are pretty independent. 

*Current progress* (updated)
In order to add inference example for the "inference project", one option is to add "infer" methods for each of the two `train_main()` and `decode_main()`. This is what is done in this PR. 
- For the infer method for `train_main`, we have to provide both source as well as target sequence, even though ideally we don't have access to the target (This is the same issue as in PR #8176). This works.
- For the infer method for `decode_main`, we add the function in Python and it works.
- The C++ unit-tests corresponding to the 2 infer methods have been added. Since, we don't have a CUDA implementation of beam search op yet, the unit-test for decode runs only for CPU.

### CODE REVIEW

File: paddle/fluid/inference/tests/book/test_inference_machine_translation_decode.cc [hunk 1]
```diff
@@ -0,0 +1,85 @@
...
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License. */
+
+#include <gtest/gtest.h>
+#include "gflags/gflags.h"
+#include "test_helper.h"
```
Comment (wangkuiyi) [inline paddle/fluid/inference/tests/book/test_inference_machine_translation_decode.cc]: Please use full path name with #include. Here it should be

```
#include "paddle/fluid/inference/tests/book/test_helper.h"
```

generic.cmake should fail if it is not full-path inclusion; I will figure out why it doesn't.

**Review by kexinzhao — APPROVED** [review]
LGTM!

File: paddle/fluid/inference/tests/book/test_inference_machine_translation_decode.cc [hunk 2]
```diff
@@ -0,0 +1,85 @@
...
+  // Run inference on CPU
+  TestInference<paddle::platform::CPUPlace>(dirname, cpu_feeds, cpu_fetchs1);
+  LOG(INFO) << result_ids1.lod();
+  LOG(INFO) << result_ids1.dims();
+  LOG(INFO) << result_scores1.lod();
+  LOG(INFO) << result_scores1.dims();
+
+#ifdef PADDLE_WITH_CUDA
+  LOG(INFO) << "Beam search isn't supported on gpu yet";
+/*
```
[the remark this answers is not in the document]
Comment (sidgoyal78) [reply paddle/fluid/inference/tests/book/test_inference_machine_translation_decode.cc]: Sure, thanks.

File: python/paddle/v2/fluid/tests/book/test_machine_translation.py [hunk 3]
```diff
@@ -29,7 +29,7 @@
 max_length = 8
 topk_size = 50
 trg_dic_size = 10000
-beam_size = 2
+beam_size = 1
```
[the remark this answers is not in the document]
Comment (sidgoyal78) [reply python/paddle/v2/fluid/tests/book/test_machine_translation.py]: I think I just changed it for debugging purposes. Thanks for pointing it out, will fix.

File: paddle/fluid/inference/tests/book/test_inference_machine_translation_decode.cc [hunk 4]
```diff
@@ -0,0 +1,85 @@
...
+
+  std::vector<int64_t> init_ids_data = {1};
+  SetupLoDTensor<int64_t>(init_ids, {1, 1}, init_lod, init_ids_data);
+
+  std::vector<float> init_scores_data = {1};
+  SetupLoDTensor<float>(init_scores, {1, 1}, init_lod, init_scores_data);
+
+  // Setup input sequence of 5 words
+  paddle::framework::LoDTensor input_sequence;
+  std::vector<int64_t> inp_data = {100, 50, 8, 94, 122};
```
Comment (Xreki) [inline paddle/fluid/inference/tests/book/test_inference_machine_translation_decode.cc]: Does this data have a specific meaning? Or can we use random data?

Comment (sidgoyal78) [reply paddle/fluid/inference/tests/book/test_inference_machine_translation_decode.cc]: This is a random sequence of 5 words (The ids need to be within a range).

**Comment by sidgoyal78** [comment]
Since there is an issue with the `test_rnn_encoder_decoder`, (https://github.com/PaddlePaddle/Paddle/blob/develop/paddle/fluid/inference/tests/book/CMakeLists.txt), and this unit tests in the PR also has a similar test. I think maybe it is better to not merge the PR yet?

File: paddle/fluid/inference/tests/book/CMakeLists.txt [hunk 5]
```diff
@@ -32,3 +32,17 @@ inference_test(recommender_system)
...
+    DEPS ARCHIVE_START paddle_fluid ARCHIVE_END
+    ARGS --dirname=${PYTHON_TESTS_DIR}/book/machine_translation_train.inference.model)
+cc_test(test_inference_machine_translation_decode
+    SRCS test_inference_machine_translation_decode.cc
+    DEPS ARCHIVE_START paddle_fluid ARCHIVE_END
+    ARGS --dirname=${PYTHON_TESTS_DIR}/book/machine_translation_decode.inference.model)
+set_tests_properties(test_inference_machine_translation_train
+    PROPERTIES DEPENDS test_machine_translation)
+set_tests_properties(test_inference_machine_translation_decode
+    PROPERTIES DEPENDS test_machine_translation)
```
[the remark this answers is not in the document]
Comment (sidgoyal78) [reply paddle/fluid/inference/tests/book/CMakeLists.txt]: I couldn't get both the tests to be expressed with the `inference_test` function. So I added explicit statements. Do you suggest modifying the `inference_test` function to accommodate this?

**Comment by sidgoyal78** [comment]
Closing the PR

**Pull request closed without merging**